### DISCUSSION TITLE
Optimise server response

### DISCUSSION THREAD

**Discussion by DavidGilWeb** [filamentphp/filament#13353] [Help]
### Package

Table builder

### Package Version

v3.2.2

### How can we help you?

Hi, I have a project made with filament and I would like to improve the speed of the server response.
I have been reading and have managed to improve it, but I think it still needs to be improved.
On a page that loads 100 rows of projects, my load is 1.40sec. 
This table draws data from two bd tables, also has a drop-down menu with 6 customisable actions. 
What could I do to try to improve this load?

**Comment by leandrocfe** [comment]
You could share some code or share a mini repo on github to reproduce this.

**Reply by DavidGilWeb** [reply]
public static function table(Table $table): Table
    {
        return $table
            ->recordClasses(fn (Model $Project) => match ($Project->Status) {
                '0' => 'opacity-30',
                '1' => 'border-s-2 border-orange-600 dark:border-orange-300',
                '2' => 'border-s-2 border-green-600 dark:border-green-300',
                default => null,
            })
            ->paginated([100, 'all'])
            ->poll('10s')
            ->defaultPaginationPageOption(100)
            ->striped()
            ->columns([
                Split::make([

                    TextColumn::make('created_at')->label('Creado')->dateTime('d/m')->sortable()->toggleable()->grow(false),
                
                    Split::make([
                        
                        TextColumn::make('clientId')->label('Cliente')->limit(20)->searchable()/*->grow(false)*/->weight(FontWeight::Bold)
                        ->wrap()->formatStateUsing(function(Project $record) : String{
                            $client = Client::select('BusinessName','Phone')->find($record->clientId);
                            $record->Phone = $client->Phone;
                            return $client->BusinessName;
                        }),
                        TextColumn::make('City')->label('Ciudad')->searchable()->toggleable()->grow(false)->formatStateUsing(function(Project $record) : String{
                            $city = isset($record->CityOverride) ? $record->CityOverride : $record->City;
                            return $city;
                        }),
                        TextColumn::make('Address')->label('Dirección')->limit(50)->weight(FontWeight::Bold)
                            ->searchable()->grow(true),

                    ])->from('max-md')->grow(true),
                    
                    TextColumn::make('BuiltSqm')->alignment(Alignment::End)->label('Sup.')->toggleable()->visibleFrom('md')->grow(false)->suffix('m²'),
                    
                    Split::make([
                        TextColumn::make('Status')->label('Estado')->wrap()->formatStateUsing(function(Project $record) : HtmlString{
                            switch ($record->Status) {
                                case 0:
                                    return new HtmlString('<span class="statusBadge status_Presupuestado" title="Esperar aprobación de cliente">Presupuestado</span>');
                              ....
                        }),

                        TextColumn::make('DateSessionScheduled')->label('Sesión')->dateTime('D d/m H:i')->sortable()->toggleable()->placeholder('--')->alignment(Alignment::Center)->weight(FontWeight::Bold),

                    ])->from('max-sm'),
                    TextColumn::make('EstMarketValue')->alignment(Alignment::End)->label('Val. Mercado')->numeric()->toggleable()->visibleFrom('lg')/*->grow(false)*/->suffix(' €'),
                    TextColumn::make('ProjectID')->label('ID')
                        /*->extraAttributes(['class' => 'copyable'])
                        ->copyable()
                        ->copyMessage('Copiado!')
                        ->copyMessageDuration(1500)*/
                        ->searchable()
                        ->toggleable()
                        ->grow(false)
                        ->visibleFrom('lg'),
                        
                    
                    TextColumn::make('PriceNetAmount')->alignment(Alignment::End)->label('Total')->money('EUR')->toggleable()->visibleFrom('md')/*->grow(false)*/->alignment(Alignment::End),
                
                ])->from('sm'),

            ])
            ->filters([
                //
            ])
            ->actions([
                
                    Tables\Actions\ActionGroup::make([
                        //Tables\Actions\EditAction::make('Modificar')->label('Modificar'),
                    Tables\Actions\ActionGroup::make([
                        Tables\Actions\Action::make('sendQuote')->label('Enviar presupuesto')->icon('heroicon-o-envelope')->color('success')
                        ->requiresConfirmation()
                        ->url(fn(Project $record)=>route('project.prevMail', ['folder' => 'emails', 'version' => $record->VersEmails, 'lang' => $record->Lang == 0 ? 'ES' : 'EN', 'template' => 'SendingQuote', 'data' => $record]))->openUrlInNewTab()

                    ])->dropdown(false),

                    Tables\Actions\ActionGroup::make([
                        Tables\Actions\Action::make('viewPdf')->label('Ver presupuesto')->icon('heroicon-o-eye')
                        ->url(fn(Project $record)=>route('project.view', ['folder' => 'quotes', 'version' => $record->VersEmails, 'lang' => $record->Lang == 0 ? 'ES' : 'EN', 'template' => 'Quotes', 'data' => $record]))->openUrlInNewTab()/* abrir con target blank */,
                        Tables\Actions\Action::make('downloadProject')->label('Descargar presupuesto PDF')->icon('heroicon-o-arrow-down-tray')
                        ->url(fn(Project $record)=>route('project.pdf.download', ['folder' => 'quotes', 'version' => $record->VersEmails, 'lang' => $record->Lang == 0 ? 'ES' : 'EN', 'template' => 'Quotes', 'data' => $record])),
                        Tables\Actions\Action::make('Duplicar proyecto') // Puedes usar cualquier icono que prefieras
                        ->requiresConfirmation() // Opcional: solicita confirmación antes de duplicar
                        ->action(function (Project $record) {
                            $GlobalFunctions = new GlobalProjectsFunctions;
                            $CSVFunction = new CSVFunctions;
                            $projectId = (string) $GlobalFunctions->createProjectId(ProjectResource::$maxLengthHexa);
                            $record->ProjectID = $projectId;
                            ....
                            $newRecord->save();
                            redirect('/projects/'.$newRecord->id.'/edit');
                        })->icon('heroicon-o-document-duplicate'),
                    ])->dropdown(false),

                    
                    Tables\Actions\ActionGroup::make([
                        Tables\Actions\Action::make('send1Pay')->label('Enviar Datos Anticipo')->icon('heroicon-o-envelope')->color('primary')->requiresConfirmation()
                        ->url(fn(Project $record)=>route('project.prevMail', ['folder' => 'emails', 'version' => $record->VersEmails, 'lang' => $record->Lang == 0 ? 'ES' : 'EN', 'template' => 'Project1stPaymentData', 'data' => $record]))->openUrlInNewTab(),
                        Tables\Actions\Action::make('sendSessionDate')->label('Enviar Fecha de Sesión')->icon('heroicon-o-envelope')->color('primary')->requiresConfirmation()
                        ->url(fn(Project $record)=>route('project.prevMail', ['folder' => 'emails', 'version' => $record->VersEmails, 'lang' => $record->Lang == 0 ? 'ES' : 'EN', 'template' => 'ProjectDateScheduled', 'data' => $record]))->openUrlInNewTab(),
                        Tables\Actions\Action::make('sendReSessionDate')->label('Enviar Reprogramación de Sesión')->icon('heroicon-o-envelope')->color('primary')->requiresConfirmation()
                        ->url(fn(Project $record)=>route('project.prevMail', ['folder' => 'emails', 'version' => $record->VersEmails, 'lang' => $record->Lang == 0 ? 'ES' : 'EN', 'template' => 'ProjectDateRescheduled', 'data' => $record]))->openUrlInNewTab(),
                        Tables\Actions\Action::make('sessionDone')->label('Marcar como Sesión Realizada')->icon('heroicon-o-check')->color('success')
                        ->action(fn(Project $record)=> $record->update(['Status' => 4])),
                        Tables\Actions\Action::make('DigitalFinish')->label('Enviar Material Digital Terminado')->icon('heroicon-o-envelope')->color('primary')->requiresConfirmation()
                        ->url(fn(Project $record)=>route('project.prevMail', ['folder' => 'emails', 'version' => $record->VersEmails, 'lang' => $record->Lang == 0 ? 'ES' : 'EN', 'template' => 'ProjectMaterialDelivery', 'data' => $record]))->openUrlInNewTab(),
                        Tables\Actions\Action::make('send2Pay')->label('Enviar Datos Saldo')->icon('heroicon-o-envelope')->color('primary')->requiresConfirmation()
                        ->url(fn(Project $record)=>route('project.prevMail', ['folder' => 'emails', 'version' => $record->VersEmails, 'lang' => $record->Lang == 0 ? 'ES' : 'EN', 'template' => 'Project2ndPaymentData', 'data' => $record]))->openUrlInNewTab()
                    ])->dropdown(false),

                    Tables\Actions\ActionGroup::make([
                        
                        Tables\Actions\Action::make('GmapsDelivery')->label('Abrir gMaps de sesión')->icon('heroicon-o-map')->color('info')->url(fn(Project $record)=> $record->AddressGoogleMapsURL)->openUrlInNewTab()
                        ->visible(fn ($record) => $record->AddressGoogleMapsURL!=null),
                        Tables\Actions\Action::make('callClient')->label('Llamar a cliente')->icon('heroicon-o-phone')->color('info')->url(fn(Project $record)=> 'tel:'.$record->Phone)->openUrlInNewTab(),
                        Tables\Actions\Action::make('whatsappClient')->label('Whatsapp a cliente')->icon('heroicon-o-chat-bubble-oval-left-ellipsis')->color('info')
                        ->url(fn(Project $record)=> 'https://wa.me/'.$record->Phone)
                        ->openUrlInNewTab()
                        //->visible(fn ($record) => Client::find($record->clientId)->Phone && (substr(Client::find($record->clientId)->Phone, 0, 1) == '+346')),
                        
                    ])->dropdown(false),

                    Tables\Actions\DeleteAction::make()->label('Eliminar proyecto')->icon('heroicon-o-trash')->requiresConfirmation()->successRedirectUrl(function(Project $record){
                        return route('deleteEvent', ['id' => $record->id]);
                    }),
                ])->dropdownWidth(MaxWidth::ExtraSmall)
            ])
            ->emptyStateActions([
                Tables\Actions\CreateAction::make(),
            ])
            ->defaultSort('created_at', 'desc');
    }

**Reply by DavidGilWeb** [reply]
This is a code snippet from the table, I cannot share repository because I have private repository with sensitive data. I hope this code helps.
